File: 프로그래머스/2/178870. 연속된 부분 수열의 합/연속된 부분 수열의 합.cpp
#include <string>
#include <vector>

using namespace std;

vector<int> solution(vector<int> sequence, int k) {
    vector<int> answer(2);
    
    int min = 1000001;
    
    int left = 0;
    int right = 0;
    int sum = sequence[0];
    
    while(right < sequence.size() || left < right)
    {
        if (sum < k)
        {
            if (right + 1 >= sequence.size())
                break;
            
            sum += sequence[++right];
        }
        else if(sum == k)
        {
            if (right - left < min)
            {
                answer[0] = left;
                answer[1] = right;
                min = right - left;
            }
            sum -= sequence[left];
            ++left;
        }
        else
        {
            sum -= sequence[left];
            ++left;
        }
    }
    
    return answer;
}

File: 프로그래머스/2/178870. 연속된 부분 수열의 합/연속된 부분 수열의 합.java
class Solution {
    public int[] solution(int[] sequence, int k) {
        int[] answer = new int[2];
        answer[1] = 1_000_001;

        int left = 0;
        int right = 0;
        int sum = 0;

        while (right < sequence.length) {

            sum += sequence[right];

            while (left <= right && sum >= k) {
                
                if(sum == k) {
                    int len = right - left;
                    if (len < answer[1] - answer[0]) {
                        answer[0] = left;
                        answer[1] = right;
                    }
                }
                sum -= sequence[left++];
            }

            right++;
        }

        
        return answer;
    }
}